src/Hyperfund.sol

```diff
@@ -75,6 +72,34 @@ contract Hyperfund is AccessControl, Pausable {
         _mintFraction(msg.sender, units);
     }
 
+    function nonfinancialContribution(address _contributor, uint256 _units)
+        external
+        whenNotPaused
+        onlyRole(MANAGER_ROLE)
+    {
+        require(_contributor != address(0), "invalid contributor");
+        require(_units != 0, "invalid units");
+        require(hypercertMinter.unitsOf(hypercertId) >= _units, "amount accedes available supply");
+        nonfinancialContributions[_contributor] += _units;
+        _mintFraction(_contributor, _units);
+    }
+
+    /// @notice redeem a hypercert fraction for the corresponding amount of tokens
+    /// NOTE: sender must first approve the hyperfund to transfer the hypercert fraction, by calling hypercertMinter.setApprovalForAll(address(this), true)
+    /// @param _fractionId id of the hypercert fraction
+    /// @param _token address of the token to redeem, must be allowlisted. address(0) for native token
+    function redeem(uint256 _fractionId, address _token) external whenNotPaused {
```

Comment:
I think we should also keep the mapping of fractionId particular address can redeem otherwise contributor create any random hypercert and retire it for underlying tokens while my allocated hypercert will be staked and accrue rewards.

Reply:
right! i added a check that the fraction is of the right hypercert. i'm not sure we should keep track of the fraction ids though, because devs can sell and buy their fractions, and i think they should be able to redeem as much as they have allowance.. wdyt?

Reply:
Ya having the check is sufficient.